### PR TITLE
Added type validation to Assertions.

### PR DESCRIPTION
**Pull request by traceyyoshima**
Changes:
- Kotlin test harness will validate types.

fixes https://github.com/openrewrite/rewrite-kotlin/issues/482
fixes https://github.com/openrewrite/rewrite-kotlin/issues/474

### CODE REVIEW

**Comment by traceyyoshima** [comment]
Blocked by https://github.com/openrewrite/rewrite-kotlin/issues/480.

File: src/main/java/org/openrewrite/kotlin/Assertions.java [hunk 1]
```diff
@@ -196,6 +209,36 @@ public static UncheckedConsumer<SourceSpec<?>> spaceConscious() {
         };
     }
 
+    private static void assertValidTypes(TypeValidation typeValidation, J sf) {
```
Comment (traceyyoshima) [inline src/main/java/org/openrewrite/kotlin/Assertions.java:212]: @knutwannheden @kunli2, I don't think mixins will help us customize the `FindMissingTypes` recipe. Initially, I thought about creating a `FindMissingKotlinTypes` recipe/visitor, which could arguably be helpful. But I recall Jonathan preferring nearly identical recipes not to be created. Eventually, if we find there is a need for the augmented recipe, we can move the code.

What do you think? Is it fine to have the visitor in Assertions for now or would you prefer a recipe?

File: src/test/java/org/openrewrite/kotlin/tree/AnnotationTest.java [hunk 2]
```diff
@@ -15,48 +15,23 @@
...
+import org.openrewrite.test.TypeValidation;
 
 import java.util.List;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.openrewrite.kotlin.Assertions.kotlin;
 
-@SuppressWarnings({"RedundantSuppression", "RedundantNullableReturnType", "RedundantVisibilityModifier", "UnusedReceiverParameter", "SortModifiers", "TrailingComma"})
+@SuppressWarnings({"RedundantSuppression", "RedundantNullableReturnType", "RedundantVisibilityModifier", "UnusedReceiverParameter", "SortModifiers", "TrailingComma", "RedundantGetter", "RedundantSetter"})
 class AnnotationTest implements RewriteTest {
```
Comment (traceyyoshima) [inline src/test/java/org/openrewrite/kotlin/tree/AnnotationTest.java:33]: I cleaned up the errors/warnings here when I looked through the tests.

**Comment by traceyyoshima** [comment]
@knutwannheden type validation is essentially ready for review. The failing tests are identifiers with aliases, and I'm looking into how to detect aliases.

```kotlin
import java.util.regex.Pattern.CASE_INSENSITIVE as i

class A {
    val f = arrayOf(i)
}
```

File: src/main/java/org/openrewrite/kotlin/Assertions.java (hunk 1 above)
Comment (knutwannheden) [reply src/main/java/org/openrewrite/kotlin/Assertions.java:212]: At some point I would be interested in the details regarding the mixins, do that we can try to come up with a fix or another approach for this. An alternative might be delegation kind of like we do in the printer.

For now I think what you currently have is good. Having type validations working is really a step forward. Thanks for fixing this!

**Review by knutwannheden — APPROVED** [review]
As we can disable validations on a test case basis, we can fix any remaining issues in follow up PRs.

**Comment by traceyyoshima** [comment]
> As we can disable validations on a test case basis, we can fix any remaining issues in follow up PRs.

Opened: https://github.com/openrewrite/rewrite-kotlin/issues/545 to track a fix for aliases. Tracking the declarations of a name will allow us to apply the appropriate field type comparison.

**Pull request merged**